tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc

```diff
@@ -4682,63 +4682,97 @@ TEST_F(OpConverterTest, ConvertConv2DBackpropInput) {
-    auto input = ops::Placeholder(s.WithOpName("input"), DT_FLOAT);
-    auto filter = ops::Placeholder(s.WithOpName("weights"), DT_FLOAT);
+// Get the NodeDef for Pack.
+NodeDef GetConv3DNodeDef(std::vector<int> strides = {1, 1, 1, 1, 1},
+                         string padding = "SAME", string data_format = "NCDHW",
+                         std::vector<int> dilations = {1, 1, 1, 1, 1},
+                         bool is_conv3d_backprop_input = false) {
+  Scope s = Scope::NewRootScope();
+  auto input = ops::Placeholder(s.WithOpName("input"), DT_FLOAT);
+  auto filter = ops::Placeholder(s.WithOpName("weights"), DT_FLOAT);
+
+  if (is_conv3d_backprop_input) {
+    auto input_sizes =
+        ops::Placeholder(s.WithOpName("input_sizes"), DT_INT32);
+    ops::Conv3DBackpropInputV2::Attrs attrs =
+        ops::Conv3DBackpropInputV2::Attrs()
+            .DataFormat(data_format)
+            .Dilations(dilations);
+    auto conv3d =
+        ops::Conv3DBackpropInputV2(s.WithOpName("my_conv3d"), input_sizes,
+                                   filter, input, strides, padding, attrs);
+    return conv3d.operation.node()->def();
+  } else {
+    ops::Conv3D::Attrs attrs =
+        ops::Conv3D::Attrs().DataFormat(data_format).Dilations(dilations);
+    auto conv3d = ops::Conv3D(s.WithOpName("my_conv3d"), input, filter,
+                              strides, padding, attrs);
+    return conv3d.operation.node()->def();
+  }
+}
 
-    if (is_conv3d_backprop_input) {
-      auto input_sizes =
-          ops::Placeholder(s.WithOpName("input_sizes"), DT_INT32);
-      ops::Conv3DBackpropInputV2::Attrs attrs =
-          ops::Conv3DBackpropInputV2::Attrs()
-              .DataFormat(data_format)
-              .Dilations(dilations);
-      auto conv3d =
-          ops::Conv3DBackpropInputV2(s.WithOpName("my_conv3d"), input_sizes,
-                                     filter, input, strides, padding, attrs);
-      return conv3d.operation.node()->def();
-    } else {
-      ops::Conv3D::Attrs attrs =
-          ops::Conv3D::Attrs().DataFormat(data_format).Dilations(dilations);
-      auto conv3d = ops::Conv3D(s.WithOpName("my_conv3d"), input, filter,
-                                strides, padding, attrs);
-      return conv3d.operation.node()->def();
-    }
-  };
+struct Conv3DTestParams {
+  std::vector<int> input_dims;
+  std::vector<float> input;
+  std::vector<int> filter_dims;
+  std::vector<float> filter;
+  std::vector<int> strides;
+  string padding;
+  string data_format;
+  std::vector<int> dilations;
+  bool is_conv3d_backprop;
+  std::vector<int> expected_output_dims;
+  std::vector<float> expected_output;
+};
+
+void TestConv3D(ParameterizedOpConverterTestBase* test, Conv3DTestParams& p) {
+  test->Reset();
+  NodeDef node_def = GetConv3DNodeDef(p.strides, p.padding, p.data_format,
+                                      p.dilations, p.is_conv3d_backprop);
 
+  test->AddTestTensor("input", p.input_dims, test->get_tf_type(), p.input);
```

Comment:
Does this test work in dynamic shape mode? AFAIK TRT requires static (known) values for the channel dimension. By default AddTestTensor replaces all input dims with -1, therefore this should fail the test in dynamic shape mode.

For the Conv2D test, I needed to fix the input dims like [this](https://github.com/tensorflow/tensorflow/blob/6fd7d69823a6cca1a83bff1db3bd34f7a7660ee4/tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc#L4538-L4543), and define an extra test for the condition on the [channel value ](https://github.com/tensorflow/tensorflow/blob/6fd7d69823a6cca1a83bff1db3bd34f7a7660ee4/tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc#L4436-L4447).

Reply:
It does work so far on my end. Let me know if you have any issue trying to reproduce this behavior

Reply:
Thanks Jonathan, you are right, it does work. Unfortunately, I am also right: in general it should not work in dynamic shape mode because TRT requires that the channel dimension is known. The following happens here: TRT is clever enough to deduce that setting `min=max` shape optimization profile means that the channel dims is known at engine creation time, and treats it as if the corresponding input dim was set to that known value (instead of -1 that we provide).

If you define a profile where min<max for the channel dimension, then TRT gives the following error message:
```
DefaultLogger /my_conv3d-conv:CONVOLUTION: number of channels in input tensor to a convolution layer must not be dynamic
DefaultLogger Builder failed while analyzing shapes.
```
I recommend the following:
1. In the converter, check that the [channel is known](https://github.com/tensorflow/tensorflow/blob/84377fe862a0bf36bb3c27166b766bc1097954ae/tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc#L2331-L2336). We can relax this a little bit: allow conversion if `kOptimal` or `kImplicitBatchModeCompatible` profiles are used, because they have min=max for non batch dimension, reject converting the op otherwise.
2. Adapt the test use give [concrete channel dimensions](https://github.com/tensorflow/tensorflow/blob/6fd7d69823a6cca1a83bff1db3bd34f7a7660ee4/tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc#L4538-L4543).

Most importantly, I think we should adapt the unit test to check that the converters work correctly for a profile where min < max. This will be a general case in the future, and we want to be prepared for that. One option would be the following:

3. Rebase the PR so that it contains the profile management logic from #45588

4. Define a `ProfileStrategy::kUnitTest` profile by setting min = opt = input_shape, max = min+1. Here is a sketch:
```c++
void SetUnitTestProfile(
    const std::vector<nvinfer1::Dims>& dimvec, std::vector<nvinfer1::Dims>* min,
    std::vector<nvinfer1::Dims>* opt, std::vector<nvinfer1::Dims>* max) {
  *min = dimvec;
  *opt = dimvec;
  *max = dimvec;
  for (auto& dim : *max) {
    for (int k = 0; k < dim.nbDims; k++) dim.d[k]++;
  }
}
```
This would come into `trt_shape_optimization_profiles.cc`, and a new case should be added [switch statement here](https://github.com/tensorflow/tensorflow/blob/132437408620c947aaa43db31ce442a2b30dec12/tensorflow/compiler/tf2tensorrt/utils/trt_shape_optimization_profiles.cc#L108).
Make the kUnitTest profile as a default profile for unit tests. (Currently it is set [here](https://github.com/tensorflow/tensorflow/blob/132437408620c947aaa43db31ce442a2b30dec12/tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc#L1479).)

5. Check if the [correct error is returned](https://github.com/tensorflow/tensorflow/blob/6fd7d69823a6cca1a83bff1db3bd34f7a7660ee4/tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc#L4436-L4447). 

Step 3-4 could be implemented in a separate PR.

My motivation to introduce a specific profile for unit test is the following:
- `BuildAndRun` sees a [single input shape](https://github.com/tensorflow/tensorflow/blob/132437408620c947aaa43db31ce442a2b30dec12/tensorflow/compiler/tf2tensorrt/convert/convert_nodes_test.cc#L1482) for the network.
- From this single shape we want to create a non-trivial profile (min!=max). `kImplicitBatchModeCompatible` only changes the batch dim, but we want to a profile that also changes other dims.
- Other profile options that we were considering for future implementations are derived from multiple input shapes, therefore they are not applicable.

@bixia1 please let us know if you have other ideas / preferences.

Reply:
I am fine with putting step 3-4 in a separated PR. However, instead of adding kUnitTest strategy, I suggest you to open a backdoor in TrtShapeOptimizationProfile, such as using a special test_only_field to indicate the need to modify the kOptimal mode,  to serve similar purpose (similar to what we have in the V2 converter python class to support the switch of implicit/explicit batch mode). The  ProfileStrategy enums will be hoisted out to a bigger scope to support the API change for dynamic shape. I don't feel that I can get pass the google code interview if we have kUnitTest in the enums.

Reply:
@tfeher I applied your change request: 1, 2 and 5: 38be62c4bbaefd83396bca72fc3bb430c52f0e04

Can you please check and confirm that works for you.
If it does, I'll squash my commits

Reply:
Thanks @DEKHTIARJonathan for the update, it looks good to me.